crates/wasi/src/preview2/pipe.rs

```diff
@@ -7,227 +7,835 @@
+                return Ok((bytes, return_state));
+            }
+            Some(Err(e)) => return Err(e.into()),
+            None => {}
+        }
 
-    async fn skip(&mut self, nelem: u64) -> Result<(u64, bool), Error> {
-        let num = io::copy(
-            &mut io::Read::take(&mut *self.borrow(), nelem),
-            &mut io::sink(),
-        )?;
-        Ok((num, num < nelem))
+        match self.receiver.try_recv() {
+            Ok(Ok((mut bytes, state))) => {
+                self.state = state;
+
+                let len = bytes.len().min(size);
+                let rest = bytes.split_off(len);
+                let return_state = if !rest.is_empty() {
+                    self.buffer = Some(Ok(rest));
+                    StreamState::Open
+                } else {
+                    self.state
+                };
+
+                Ok((bytes, return_state))
+            }
+            Ok(Err(e)) => Err(e.into()),
+            Err(TryRecvError::Empty) => Ok((Bytes::new(), self.state)),
+            Err(TryRecvError::Disconnected) => Err(anyhow::anyhow!(
+                "AsyncReadStream sender died - should be impossible"
+            )),
+        }
     }
 
-    async fn readable(&self) -> Result<(), Error> {
+    async fn ready(&mut self) -> Result<(), Error> {
+        if self.buffer.is_some() || self.state == StreamState::Closed {
+            return Ok(());
+        }
+        match self.receiver.recv().await {
+            Some(Ok((bytes, state))) => {
+                if state == StreamState::Closed {
+                    self.state = state;
+                }
+                self.buffer = Some(Ok(bytes));
+            }
+            Some(Err(e)) => self.buffer = Some(Err(e)),
+            None => {
+                return Err(anyhow::anyhow!(
+                    "no more sender for an open AsyncReadStream - should be impossible"
+                ))
+            }
+        }
         Ok(())
     }
 }
 
-/// A virtual pipe write end.
-///
-/// ```no_run
-/// use wasmtime_wasi::preview2::{pipe::WritePipe, WasiCtxBuilder, Table};
-/// let mut table = Table::new();
-/// let stdout = WritePipe::new_in_memory();
-/// let mut ctx = WasiCtxBuilder::new().set_stdout(stdout.clone()).build(&mut table).unwrap();
-/// // use ctx and table in an instance, then make sure it is dropped:
-/// drop(ctx);
-/// drop(table);
-/// let contents: Vec<u8> = stdout.try_into_inner().expect("sole remaining reference to WritePipe").into_inner();
-/// println!("contents of stdout: {:?}", contents);
-/// ```
 #[derive(Debug)]
-pub struct WritePipe<W: Write> {
-    writer: Arc<RwLock<W>>,
+enum WriteState {
+    Ready,
+    Pending,
+    Err(std::io::Error),
 }
 
-impl<W: Write> Clone for WritePipe<W> {
-    fn clone(&self) -> Self {
-        Self {
-            writer: self.writer.clone(),
-        }
-    }
+/// Provides a [`HostOutputStream`] impl from a [`tokio::io::AsyncWrite`] impl
+pub struct AsyncWriteStream {
+    state: Option<WriteState>,
```

Comment:
Instead of an `Option` here perhaps add `Closed` to `WriteState`?

Reply:
we had it factored that way, then we were taking advantage of Option::take() so changing the repr made sense, but now I think I can switch it back...